flake: `controllers_test.go`

Seen on `main` in https://github.com/coder/coder/actions/runs/11900586976/job/33161756273
Looks to be caused by `pipeDialer`, if the `Dial` function is called after the test has finished, one of the dial steps will somehow fail, and cause the no error assertion to fail. (The whole function looks self-contained, so I'm not sure how this is happening)

```
=== FAIL: tailnet TestInMemoryCoordination (unknown)
panic: Fail in goroutine after TestController_Disconnects has completed

goroutine 246 [running]:
testing.(*common).Fail(0xc0004c2b60)
	/opt/hostedtoolcache/go/1.22.8/x64/src/testing/testing.go:952 +0xd4
testing.(*common).Errorf(0xc0004c2b60, {0x1a97734?, 0x411c25?}, {0xc00069e250?, 0x1751ba0?, 0x101?})
	/opt/hostedtoolcache/go/1.22.8/x64/src/testing/testing.go:1075 +0x5e
github.com/stretchr/testify/assert.Fail({0x1d435c0, 0xc0004c2b60}, {0xc000646060, 0x2b}, {0x0, 0x0, 0x0})
	/home/runner/go/pkg/mod/github.com/stretchr/testify@v1.9.0/assert/assertions.go:363 +0x370
github.com/stretchr/testify/assert.NoError({0x1d435c0, 0xc0004c2b60}, {0x1d43500, 0xc0002dedc0}, {0x0, 0x0, 0x0})
	/home/runner/go/pkg/mod/github.com/stretchr/testify@v1.9.0/assert/assertions.go:1545 +0x125
github.com/coder/coder/v2/tailnet_test.(*pipeDialer).Dial(0xc000143800, {0xc0000ba230?, 0x1?}, {0x1aca93d?, 0x1f?})
	/home/runner/work/coder/coder/tailnet/controllers_test.go:1223 +0x27c
github.com/coder/coder/v2/tailnet.(*Controller).Run.func1()
	/home/runner/work/coder/coder/tailnet/controllers.go:1166 +0x139
created by github.com/coder/coder/v2/tailnet.(*Controller).Run in goroutine 84
	/home/runner/work/coder/coder/tailnet/controllers.go:1161 +0x109
```

--- a/tailnet/controllers.go
+++ b/tailnet/controllers.go
@@ -1163,10 +1163,17 @@ func (c *Controller) Run(ctx context.Context) {
 		// Sadly retry doesn't support quartz.Clock yet so this is not
 		// influenced by the configured clock.
 		for retrier := retry.New(50*time.Millisecond, 10*time.Second); retrier.Wait(c.ctx); {
+			// Check the context again before dialing, since `retrier.Wait()` could return true
+			// if the delay is 0, even if the context was canceled. This ensures we don't redial
+			// after a graceful shutdown.
+			if c.ctx.Err() != nil {
+				return
+			}
+
 			tailnetClients, err := c.Dialer.Dial(c.ctx, c.ResumeTokenCtrl)
 			if err != nil {
-				if xerrors.Is(err, context.Canceled) {
-					continue
+				if xerrors.Is(err, context.Canceled) || xerrors.Is(err, context.DeadlineExceeded) {
+					return
 				}
 				errF := slog.Error(err)
 				var sdkErr *codersdk.Error

--- a/tailnet/controllers_test.go
+++ b/tailnet/controllers_test.go
@@ -14,7 +14,6 @@ import (
 
 	"github.com/google/uuid"
 	"github.com/hashicorp/yamux"
-	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 	"go.uber.org/mock/gomock"
 	"golang.org/x/xerrors"
@@ -1071,6 +1070,7 @@ func TestController_Disconnects(t *testing.T) {
 	close(call.Resps)
 
 	_ = testutil.RequireRecvCtx(testCtx, t, peersLost)
+	_ = testutil.RequireRecvCtx(testCtx, t, uut.Closed())
 }
 
 func TestController_TelemetrySuccess(t *testing.T) {
@@ -1210,24 +1210,28 @@ type pipeDialer struct {
 
 func (p *pipeDialer) Dial(_ context.Context, _ tailnet.ResumeTokenController) (tailnet.ControlProtocolClients, error) {
 	s, c := net.Pipe()
+	p.t.Cleanup(func() {
+		_ = s.Close()
+		_ = c.Close()
+	})
 	go func() {
 		err := p.svc.ServeConnV2(p.ctx, s, p.streamID)
 		p.logger.Debug(p.ctx, "piped tailnet service complete", slog.Error(err))
 	}()
 	client, err := tailnet.NewDRPCClient(c, p.logger)
-	if !assert.NoError(p.t, err) {
+	if err != nil {
 		_ = c.Close()
 		return tailnet.ControlProtocolClients{}, err
 	}
 	coord, err := client.Coordinate(context.Background())
-	if !assert.NoError(p.t, err) {
+	if err != nil {
 		_ = c.Close()
 		return tailnet.ControlProtocolClients{}, err
 	}
 
 	derps := &tailnet.DERPFromDRPCWrapper{}
 	derps.Client, err = client.StreamDERPMaps(context.Background(), &proto.StreamDERPMapsRequest{})
-	if !assert.NoError(p.t, err) {
+	if err != nil {
 		_ = c.Close()
 		return tailnet.ControlProtocolClients{}, err
 	}

